rio_tiler_mvt/mvt.pyx

```diff
@@ -47,9 +53,78 @@ cpdef bytes encoder(
         # TODO: fix https://github.com/tilery/python-vtzero/issues/3
         for bidx in range(data.shape[0]):
             feature.add_property(
-                band_names[bidx].encode(), 
+                band_names[bidx].encode(),
                 str(data[bidx, idx[1], idx[0]]).encode()
             )
         feature.commit()
 
-    return mvt.serialize()
\ No newline at end of file
+    return mvt.serialize()
+
+
+cpdef bytes shapes_encoder(
+    data,
+    mask,
+    str layer_name = "my_layer",
+    dict colormap = {},
+    dict class_names = {},
+):
+    """Encode polygon extracted from GDAL polygonize.
+
+    Args:
+        data (numpy.ndarray): image data.
+        mask (numpy.ndarray): mask array.
+        layer_name (str): MVT layer name.
+        colormap (dict): GDAL colormap. If provided a `color` value will be added
+            to the feature properties.
+        class_names (dict): Dictionary mapping pixel value with class names. If provided
+            a `name` value will be added to the feature properties.
+
+    Returns:
+        bytes: Mapbox Vector Tile.
+
+    """
+    cdef float x, y
+    cdef int r, g, b
+
+    mvt = Tile()
+    mvt_layer = Layer(mvt, layer_name.encode())
+
+    # scaling factor
+    cdef int sc = 4096 // max(data.shape)
+
+    # Iterate through shapes and values
+    for (p, v) in _shapes(data, mask, connectivity=4, transform=IDENTITY.scale(sc)):
+        # make sure v is the same datatype as the input data
+        v = data.dtype.type(v)
+        feature = Polygon(mvt_layer)
+        feature.set_id(v)
+
+        polygon = geometry.shape(p)
+        # Simple GEOS Validation
+        if not polygon.is_valid:
+            # If the polygon is not valid we use a little trick
+            polygon = polygon.buffer(0)
```

Comment:
seems that applying a 0 buffer creates a valid geometry (for GEOS)

Reply:
Read that the other day too, when I had a large polygonize output

Reply:
so this is still WIP. (see https://github.com/cogeotiff/rio-tiler-mvt/issues/1)

I've deployed a tiler using https://github.com/developmentseed/titiler-mvt but I'm getting a lot of warnings because the geometry are still invalid.

I may try to use pygeos instead of shapely (and maybe try wagyu again)